File: 백준/Silver/1932. 정수 삼각형/정수 삼각형.cc
#include <iostream>
#include <vector>

using namespace std;

int n, answer = 0;
vector<vector<int>> triangle;
int dp[505][505];

int main(void) {
    cin.tie(NULL); ios_base::sync_with_stdio(false);

    cin >> n;
    for (int i = 1; i <= n; i++) {
        vector<int> v;
        for (int j = 0; j < i; j++) {
            int tmp;
            cin >> tmp;
            v.push_back(tmp);
        }
        triangle.push_back(v);
    }
    
    // dp[n][i] : n번째줄 i번째 수까지의 최대 합
    dp[0][0] = triangle[0][0];
    answer = dp[0][0];
    for (int i = 1; i < n; i++) {
        for (int j = 0; j <= i; j++) {
            if (j == 0)
                dp[i][j] = dp[i - 1][0] + triangle[i][j];
            else if (j == i)
                dp[i][j] = dp[i - 1][j - 1] + triangle[i][j];
            else
                dp[i][j] = max(dp[i - 1][j - 1], dp[i - 1][j]) + triangle[i][j];
            answer = max(answer, dp[i][j]);
        }
    }
    
    cout << answer << "\n";
    
    return 0;
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
    answer = 0
    
    # dp 배열 초기화
    n = len(triangle)
    dp = [[] for _ in range(n)]
    dp[0].append(triangle[0][0])

    for i in range(1, n): # i번째 줄
        for j in range(i + 1): # j번째 수
            # 바로 윗줄 0번째 수에서만 갈 수 있음
            if j == 0:
                dp[i].append(dp[i - 1][j] + triangle[i][j])
            # 바로 윗줄 j - 1번째 수에서만 갈 수 있음
            elif j == i:
                dp[i].append(dp[i - 1][j - 1] + triangle[i][j])
            # 바로 윗줄 양쪽에서 갈 수 있음
            else:
                cost = max(dp[i - 1][j - 1], dp[i - 1][j]) + triangle[i][j]
                dp[i].append(cost)
    
    answer = max(dp[n - 1])
    
    return answer